libeos-updater-util/flatpak-util.c

```diff
@@ -1883,12 +1883,29 @@ euu_add_dependency_ref_actions_for_installation (FlatpakInstallation  *installat
                          euu_flatpak_remote_ref_action_ref (ref_action));
 
       /* Go through each of the related refs and add it to the dependency ref
-       * actions. Note that we may be adding duplicates here but they will be
-       * squashed below. */
+       * actions. Note that we may be adding duplicates here for uninstall
+       * actions but they will be squashed below. */
       for (gsize j = 0; j < related_ref_actions->len; ++j)
         {
           EuuFlatpakRemoteRefAction *related_ref_action = g_ptr_array_index (related_ref_actions, j);
 
+          /* In case multiple actions try to install/update the same
+           * dependency, we should let the first one win since it will be
+           * executed first. A side effect of this is that if two actions try
+           * to install the same dependency from different remotes then the
+           * first one wins. Can't satisfy them all, really.
```

Comment:
Is this not already handled by `squash_ref_actions_ptr_array()`? I guess not, since otherwise you wouldn’t have made this commit — but perhaps the fix lies in that function rather than this one?

Reply:
No, because `squash_ref_actions_ptr_array()` lets the last action on a ref win, and that is what we want when action A and action B affect the same ref, where both actions are "main" actions not dependencies. But when action A and action B both have a dependency to install/update C, we want the dependency from A to win. So that's why we need this patch.